### PR TITLE
[tool][android] Allow --target-platform work properly with --debug mode

### LINKED ISSUE #153359: build apk ignores --target-platform if --debug is set
### Steps to reproduce

1. `flutter create testproject`
2. `cd testproject`
3. `flutter build apk --target-platform=android-arm64 --debug`

### Expected results

The resulting `app-debug.apk` only contains code for android-arm64.

### Actual results

The resulting `app-debug.apk` contains code for android-arm, android-arm64, android-x86, android-x64.

When running without `--debug`, i.e. `flutter build apk --target-platform=android-arm64` then the resulting `app-release.apk` correctly contains only code for android-arm64.

### Why this is relevant

I am using https://github.com/cunarist/rinf. Sometimes you have to do a `flutter clean`. And (re)compiling Rust code for 4 target archs takes a loooong time. But I only need and only want to compile for one arch. I expect `--target-platform` to work with `--debug`.

Alternatively (but not ideal), the output of `flutter build apk --help` should indicate that these two options are incompatible.

### Code sample

No need for a code sample. Just run `flutter create testproject` and use that.

### Screenshots or Video

n/a

### Logs

n/a

### Flutter Doctor output

<details open><summary>Doctor output</summary>

```console
Doctor summary (to see all details, run flutter doctor -v):
[✓] Flutter (Channel stable, 3.24.0, on macOS 14.5 23F79 darwin-arm64, locale en-CH)
[✓] Android toolchain - develop for Android devices (Android SDK version 35.0.0)
[✓] Xcode - develop for iOS and macOS (Xcode 15.4)
[✗] Chrome - develop for the web (Cannot find Chrome executable at /Applications/Google Chrome.app/Contents/MacOS/Google Chrome)
    ! Cannot find Chrome. Try setting CHROME_EXECUTABLE to a Chrome executable.
[✓] Android Studio (version 2024.1)
[✓] VS Code (version 1.92.0)
[✓] Connected device (3 available)
[✓] Network resources

! Doctor found issues in 1 category
```

</details>

### PR DESCRIPTION
**Pull request by Sameri11**
This PR addresses an issue where the `--target-platform` flag was not being respected when building APKs in debug mode. Previously, debug builds would always include `x86` and `x64` architectures, regardless of the specified target platform. This change ensures that the `--target-platform` flag is honored across all build modes, including debug.

To achieve this, `BuildApkCommand` has been slightly changed to become responsible for list of archs that should be built in the current run,rather than just parsing arguments. Previously, this responsibility was distributed to gradle, which could be frustrating (in my opinion)

Fixes #153359 

## Pre-launch Checklist

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [x] I read and followed the [Flutter Style Guide], including [Features we expect every widget to implement].
- [x] I signed the [CLA].
- [x] I listed at least one issue that this PR fixes in the description above.
- [x] I updated/added relevant documentation (doc comments with `///`).
- [x] I added new tests to check the change I am making, or this PR is [test-exempt].
- [x] I followed the [breaking change policy] and added [Data Driven Fixes] where supported.
- [x] All existing and new tests are passing.

If you need help, consider asking for advice on the #hackers-new channel on [Discord].

[Contributor Guide]: https://github.com/flutter/flutter/blob/main/docs/contributing/Tree-hygiene.md#overview
[Tree Hygiene]: https://github.com/flutter/flutter/blob/main/docs/contributing/Tree-hygiene.md
[test-exempt]: https://github.com/flutter/flutter/blob/main/docs/contributing/Tree-hygiene.md#tests
[Flutter Style Guide]: https://github.com/flutter/flutter/blob/main/docs/contributing/Style-guide-for-Flutter-repo.md
[Features we expect every widget to implement]: https://github.com/flutter/flutter/blob/main/docs/contributing/Style-guide-for-Flutter-repo.md#features-we-expect-every-widget-to-implement
[CLA]: https://cla.developers.google.com/
[flutter/tests]: https://github.com/flutter/tests
[breaking change policy]: https://github.com/flutter/flutter/blob/main/docs/contributing/Tree-hygiene.md#handling-breaking-changes
[Discord]: https://github.com/flutter/flutter/blob/main/docs/contributing/Chat.md
[Data Driven Fixes]: https://github.com/flutter/flutter/blob/main/docs/contributing/Data-driven-Fixes.md

### CODE REVIEW

**Comment by Sameri11** [comment]
FYI @reidbaker based on activity in #153359

**Review by gmackall — APPROVED** [review]
This lgtm, but I'd also like to get review from a member of the tools team before merging

**Review by andrewkolos — APPROVED** [review]
LGTM with nitpicks. Thanks!

File: packages/flutter_tools/lib/src/commands/build_apk.dart
Comment (Sameri11) [reply packages/flutter_tools/lib/src/commands/build_apk.dart]: Deleted it.

**Review by reidbaker — APPROVED** [review]

File: packages/flutter_tools/lib/src/commands/build_apk.dart
Comment (Sameri11) [reply packages/flutter_tools/lib/src/commands/build_apk.dart:73]: Actually, `x86` was included into debug builds by default before these changes, as you can see here, for example: https://github.com/flutter/flutter/pull/154476/files#diff-d6576605592ef666ab7c75963cce03509e2ff46a3db735e4372754c9f9809681L37

Comment indicates that `x86` was intentionally included along with `x86_64`, and issue was about to *not* include them if `--target-platform` specified. So, I guess in this case, this behaviour was preserved: if `--target-platform` is not set – apk will contain all 4 arches (3 for release mode).

In issue you linked, I can see that, despite `flutter build apk --debug` being called, `release` variant is being configured along with `debug` in gradle. This behaviour makes me think that there's something wrong with gradle configuration in that project.
But, I can't link these changes to the issue you provided right away, will need to investigate.

Comment (martinhaintz) [reply packages/flutter_tools/lib/src/commands/build_apk.dart:73]: Thank you @Sameri11 for the fast and detailed answer. We will investigate this in our project.

Comment (romtsn) [reply packages/flutter_tools/lib/src/commands/build_apk.dart:73]: @Sameri11 that's how the android gradle plugin works, even if you call `assembleDebug` it will still run the configuration phase for all variants (that is, release and debug by default), because it has to calculate the task graph and whatnot and is variant-agnostic at this stage of the build.

If I'm reading it right, the offending change is this line: https://github.com/flutter/flutter/pull/154476/files#diff-679494ea4187d01a5fa3d1272e537bc58c1f504cd407d99718be5e1fd49b3c93L51
It used to default to the 3 platforms when `--target-platform` was not specified, but now it defaults to 4 and that is what breaks.

Comment (Sameri11) [reply packages/flutter_tools/lib/src/commands/build_apk.dart:73]: Hi @romtsn ! You're right about configuration phase, my bad. What I tried to say is that, for some reason, there is attempt to **resolve** dependencies during configuration time in getsentry example project, which is an error in many cases (but not always, though). You can see gradle warns about it if run build with `--verbose` flag.

As for "default to the 3 platforms before" – it's partially true because [this line](https://github.com/flutter/flutter/pull/154476/files#diff-679494ea4187d01a5fa3d1272e537bc58c1f504cd407d99718be5e1fd49b3c93L51) indicated that, but there was special logic in gradle build itself [here](https://github.com/flutter/flutter/pull/154476/files#diff-f0cdaf8963c97a175f64fe0fad130260d0614dd6cecd5cbe9580628bd51c5a4cL641-L644), this logic added `x86` in debug builds and, in the end, with `flutter build apk --debug` you got all 4 arches.

The line you pointing to as an offending sets default architectures for **debug** build. Further this file, there is also a list for `aot` architectures, which doesn't contain `x86` – these defaults are used for `release` builds. So `x86` was never meant to make it to `release` build.

What makes me think that error is the combination of Flutter build and [androidNativeBundle](https://github.com/howardpang/androidNativeBundle) is these lines from [issue](https://github.com/getsentry/sentry-dart/issues/2381) in getsentry:

```groovy
 > Could not resolve all dependencies for configuration ':app:releaseCompileClasspath'.
      > Could not find io.flutter:x86_release:1.0.0-af0f0d559c8a87d912a20971bbd84afc80a54b0f.
```

Question here is how `x86` even get to `releaseCompileClasspath`.

My best guess is that in previous Flutter versions, `x86` was added to desired archs **later** and inside actual gradle configuration, not before build.

it looked like this:
```groovy
       // this happens for each buildType in project, no matter which flutter build type is building now.

        List<String> platforms = getTargetPlatforms().collect() // default is android-x64, android-arm, android-arm64
        // Debug mode includes x86 and x64, which are commonly used in emulators.
        if (flutterBuildMode == "debug" && !useLocalEngine()) {
            platforms.add("android-x86")
            platforms.add("android-x64")
        }
        // `x86` was added only in debug configuration
        // If it's debug buildType, we have android-x64, android-arm, android-arm64, android-x86 – 4 deps.
        // If it's release buildType, we have android-x64, android-arm, android-arm64 – 3 deps.
        platforms.each { platform ->
            String arch = PLATFORM_ARCH_MAP[platform].replace("-", "_")
            // Add the `libflutter.so` dependency.
            addApiDependencies(project, buildType.name,
                     // Here dependencies added, 3 or 4, depending on buildType (calculated earlier).
                    "io.flutter:${arch}_$flutterBuildMode:$engineVersion")
        }
```

Now, desired architectures are known and set before gradle build, the just passed to it. It looks like this:

```groovy
       // this happens for each buildType in project, no matter which flutter build type is building now.

        List<String> platforms = getTargetPlatforms().collect() // default is android-x64, android-arm, android-arm64 AND android-x86

       // No additional logic for adding x86 in debug.
        // If it's debug buildType, we have android-x64, android-arm, android-arm64, android-x86 – 4 deps.
        // If it's release buildType, we have android-x64, android-arm, android-arm64, android-x86 – still 4 deps.
        platforms.each { platform ->
            String arch = PLATFORM_ARCH_MAP[platform].replace("-", "_")
            // Add the `libflutter.so` dependency.
            addApiDependencies(project, buildType.name,
                     // always 4 dependencies added.
                    "io.flutter:${arch}_$flutterBuildMode:$engineVersion")
        }
```

Main thing here, is that it's not an issue in standard Flutter build, because dependencies are not resolved during configuration for each variant there. And there can only be one build type building at single moment of time.

If you build with `--debug` – 4 arches added for every buildType, but only resolved for `debug` variant, which allowed to have `x86` dependency. So, `io.flutter.x86-release-....` added to release, but never resolved, because it's debug build happening now, and build works.

But, if I'm right, `androidNativeBundle` plugin tries to resolve all configurations inside it's `apply` method – in Gradle's configuration phase: https://github.com/howardpang/androidNativeBundle/blob/master/buildSrc/src/main/groovy/com/yy/android/gradle/nativedepend/NativeBundleImportPlugin.groovy#L175. And that breaks the build in configuration phase.

That said, feel free to open an issue in https://github.com/flutter/flutter/issues. I'm not in the right to decide if it'll be fixed because I'm not actual maintainer, and this behaviour is a part of 3rd party gradle plugin also. I guess, for minimal reproduction it will be enough to just create empty flutter project with android  and add [androidNativeBundle](https://github.com/howardpang/androidNativeBundle) to build the way it added to build in getsentry.

Also, @gmackall and @reidbaker might be interested in this case.

Comment (romtsn) [reply packages/flutter_tools/lib/src/commands/build_apk.dart:73]: @Sameri11 thanks for spending your time looking into this, this all makes sense!

> But, if I'm right, androidNativeBundle plugin tries to resolve all configurations inside it's apply method – in Gradle's configuration phase: [howardpang/androidNativeBundle@master/buildSrc/src/main/groovy/com/yy/android/gradle/nativedepend/NativeBundleImportPlugin.groovy#L175](https://github.com/howardpang/androidNativeBundle/blob/master/buildSrc/src/main/groovy/com/yy/android/gradle/nativedepend/NativeBundleImportPlugin.groovy#L175). And that breaks the build in configuration phase.

Yep, you're right, the plugin should switch to a task that receives the configuration dependencies as a `Provider` to prevent eager config resolution. We'll take it from here!

Comment (martinhaintz) [reply packages/flutter_tools/lib/src/commands/build_apk.dart:73]: @Sameri11 thanks also from my side for your detailed explanation. I just wanted to mention, that the maintainer of the `androidNativeBundle` released a new version `1.1.5`, which fixes the incompatibility.

**Pull request merged**
